src/main/drivers/flash_w25q128fv.c

```diff
@@ -0,0 +1,451 @@
+#define W25Q128FV_TIMEOUT_RESET_MS              2           // tRST = 30us
+#define W25Q128FV_TIMEOUT_BLOCK_ERASE_64KB_MS   2000        // tBE2max = 2000ms, tBE2typ = 150ms
+#define W25Q128FV_TIMEOUT_CHIP_ERASE_MS        (200 * 1000) // tCEmax 200s, tCEtyp = 40s
+
+#define W25Q128FV_TIMEOUT_PAGE_PROGRAM_MS       2           // tPPmax = 700us, tPPtyp = 250us
+#define W25Q128FV_TIMEOUT_WRITE_ENABLE_MS       2
+
+
+typedef enum {
+    INITIAL_MODE_SPI = 0,
+    INITIAL_MODE_QUADSPI,
+} w25q128fv_initialMode_e;
+
+typedef struct w25q128fvState_s {
+    w25q128fv_initialMode_e initialMode;
+    uint32_t currentWriteAddress;
+} w25q128fvState_t;
+
+w25q128fvState_t w25q128fvState = { 0 };
+
+static bool w25q128fv_waitForReady(flashDevice_t *fdevice);
+static void w25q128fv_waitForTimeout(flashDevice_t *fdevice);
+
+static void w25q128fv_setTimeout(flashDevice_t *fdevice, uint32_t timeoutMillis)
+{
+    uint32_t now = HAL_GetTick();
+    fdevice->timeoutAt = now + timeoutMillis;
+}
+
+static void w25q128fv_performOneByteCommand(flashDeviceIO_t *io, uint8_t command)
+{
+    QUADSPI_TypeDef *quadSpi = io->handle.quadSpi;
+    quadSpiTransmit1LINE(quadSpi, command, 0, NULL, 0);
+}
+
+static void w25q128fv_performCommandWithAddress(flashDeviceIO_t *io, uint8_t command, uint32_t address)
+{
+    QUADSPI_TypeDef *quadSpi = io->handle.quadSpi;
+
+    quadSpiInstructionWithAddress1LINE(quadSpi, command, 0, address & 0xffffff, W25Q128FV_ADDRESS_BITS);
+}
+
+static void w25q128fv_writeEnable(flashDevice_t *fdevice)
+{
+    w25q128fv_performOneByteCommand(&fdevice->io, W25Q128FV_INSTRUCTION_WRITE_ENABLE);
+}
+
+static uint8_t w25q128fv_readRegister(flashDeviceIO_t *io, uint8_t command)
+{
+    QUADSPI_TypeDef *quadSpi = io->handle.quadSpi;
+
+    uint8_t in[1];
+    quadSpiReceive1LINE(quadSpi, command, 0, in, W25Q128FV_STATUS_REGISTER_BITS / 8);
+
+    return in[0];
+}
+
+static void w25q128fv_writeRegister(flashDeviceIO_t *io, uint8_t command, uint8_t data)
+{
+    QUADSPI_TypeDef *quadSpi = io->handle.quadSpi;
+
+   quadSpiTransmit1LINE(quadSpi, command, 0, &data, W25Q128FV_STATUS_REGISTER_BITS / 8);
+}
+
+static void w25q128fv_deviceReset(flashDevice_t *fdevice)
+{
+    flashDeviceIO_t *io = &fdevice->io;
+
+    w25q128fv_waitForReady(fdevice);
+    w25q128fv_performOneByteCommand(io, W25Q128FV_INSTRUCTION_ENABLE_RESET);
+    w25q128fv_performOneByteCommand(io, W25Q128FV_INSTRUCTION_RESET_DEVICE);
+
+    w25q128fv_setTimeout(fdevice, W25Q128FV_TIMEOUT_RESET_MS);
+    w25q128fv_waitForTimeout(fdevice);
+
+    w25q128fv_waitForReady(fdevice);
+
+#ifdef DISABLE_NONVOLATILE_QE_MODE // Use this if you encounter a chip with it's QE bit enabled when it shouldn't be.
+    w25q128fv_performOneByteCommand(io, W25Q128FV_INSTRUCTION_WRITE_ENABLE);
+    w25q128fv_writeRegister(io, W25Q128FV_INSTRUCTION_WRITE_STATUS2_REG, 0x00);
+#endif
+
+
+#if defined(USE_FLASH_WRITES_USING_4LINES) || defined(USE_FLASH_READS_USING_4LINES)
+    uint8_t registerValue;
```

Comment:
There is no need to hoist these variables. Betaflight coding standard is to declare them on definition.

Reply:
can you explain what you mean by 'hoist'?   I'm not sure what to do.

Reply:
@mikeller got a reply for me?

Reply:
```uint8_t registerValue = w25q128fv_readRegister(io, W25Q128FV_INSTRUCTION_READ_STATUS2_REG);``` maybe?
```newValue``` could be declared in the ```if``` where it's used

Reply:
done, i understand the 'hoisting' term now.